[MLIR][OpenMP] Add MLIR Lowering Support for dist_schedule

Stylie777 opened the pull request:
`dist_schedule` was previously supported in Flang/Clang but was not
implemented in MLIR, instead a user would get a "not yet implemented"
error. This patch adds support for the `dist_schedule` clause to be
lowered to LLVM IR when used in an `omp.distribute` or `omp.wsloop`
section.

There has needed to be some rework required to ensure that MLIR/LLVM
emits the correct Schedule Type for the clause, as it uses a different
schedule type to other OpenMP directives/clauses in the runtime library.

This patch also ensures that when using dist_schedule or a chunked
schedule clause, the correct llvm loop parallel accesses details are
added.

Stylie777 committed:
[MLIR][OpenMP] Add support for `dist_schedule`

`dist_schedule` was previously supported in Flang/Clang but was not
implemented in MLIR, instead a user would get a "not yet implemented"
error. This patch adds support for the `dist_schedule` clause to be
lowered to LLVM IR when used in an `omp.distribute` or `omp.wsloop`
section.

There has needed to be some rework required to ensure that MLIR/LLVM
emits the correct Schedule Type for the clause, as it uses a different
schedule type to other OpenMP directives/clauses in the runtime library.

This patch also ensures that when using dist_schedule or a chunked
schedule clause, the correct llvm loop parallel accesses details are
added.

tblah reviewed (commented):
Please could you update `flang/docs/OpenMPSupport.md` to show dist_schedule is now supported.

Stylie777 commented on llvm/include/llvm/Frontend/OpenMP/OMP.td:
```diff
@@ -458,7 +458,8 @@ def OMP_SCHEDULE_Dynamic : EnumVal<"dynamic", 3, 1> {}
 def OMP_SCHEDULE_Auto : EnumVal<"auto", 5, 1> {}
 def OMP_SCHEDULE_Runtime : EnumVal<"runtime", 6, 1> {}
-def OMP_SCHEDULE_Default : EnumVal<"default", 7, 0> { let isDefault = 1; }
+def OMP_SCHEDULE_Distribute : EnumVal<"distribute", 7, 1> {}
```
Yes

Stylie777 commented:
I have converted this to a draft for the time being while I investigate some inconsistencies between MLIR generation with this patch and Clang.

Stylie777 reviewed (commented):
This PR is now ready for review again.

Stylie777 committed:
formatting and change to DocString for function

tblah reviewed (approved):
LGTM with a few nits

Stylie777 committed:
Respond to Review Comments

Stylie777 committed:
Remove unneeded Function Argument from `applyDynamicWorkshareLoop`

dist_schedule is not supported for dynamically allocated loops, so we
can remove the need for the DistScheduleChunkSize from this function.

Stylie777 committed:
Fix failing test

Stylie777 committed:
Respond to review comments

Stylie777 committed:
formatting§
